cou/apps/app.py

```diff
@@ -182,7 +276,279 @@ def _get_os_origin(self) -> str:
+
+    def post_upgrade_plan(self, plan: UpgradeStep) -> None:
+        """Post Upgrade planning.
+
+        :param plan: Plan that will add post upgrade as sub steps.
+        :type plan: UpgradeStep
+        """
+        self.reached_expected_target(plan)
+
+    def generate_upgrade_plan(self, target: Optional[str]) -> Optional[UpgradeStep]:
+        """Generate full upgrade plan for an Application.
+
+        :param target: OpenStack codename to upgrade.
+        :type target: Optional[str]
+        :return: Full upgrade plan if the Application is able to generate it.
+        :rtype: Optional[UpgradeStep]
+        """
+        if not target:
+            logger.warning("There is no target to upgrade.")
+            return None
+        target_version = OpenStackRelease(target)
+
+        if not self.can_generate_upgrade_plan():
+            logger.warning("Aborting upgrade plan for '%s'", self.name)
+            return None
+        if self.current_os_release >= target_version:
+            logger.warning(
+                "Application: '%s' already on a newer version than %s. Aborting upgrade.",
+                self.name,
+                target,
+            )
+            return None
+        upgrade_plan = UpgradeStep(
+            description=(
+                f"Upgrade plan for '{self.name}' from: {self.current_os_release} "
+                f"to {self.next_os_release}"
+            ),
+            parallel=False,
+            function=None,
+        )
+        self.pre_upgrade_plan(upgrade_plan)
+        self.upgrade_plan(upgrade_plan)
+        self.post_upgrade_plan(upgrade_plan)
+        return upgrade_plan
+
+    def refresh_current_channel(self, plan: UpgradeStep, parallel: bool = False) -> None:
+        """Add Plan for refresh current channel.
+
+        This function also identify if charm comes from charmstore and in that case,
+        makes the migration.
+        :param plan: Plan to add refresh current channel as sub step.
+        :type plan: UpgradeStep
+        :param parallel: Parallel running, defaults to False
+        :type parallel: bool
+        """
+        switch = None
+        description = (
+            f"Refresh '{self.name}' to the latest revision of '{self.expected_current_channel}'"
+        )
+
+        if self.channel == self.next_channel:
+            logger.warning(
+                "App: %s already has the channel set for the next OpenStack version %s",
+                self.name,
+                self.next_os_release,
+            )
+            return
+
+        if self.charm_origin == "cs":
+            description = f"Migration of '{self.name}' from charmstore to charmhub"
+            switch = f"ch:{self.charm}"
+
+        elif self.channel not in (self.expected_current_channel, self.next_channel):
+            description = (
+                f"Changing '{self.name}' channel from: '{self.channel}' "
+                f"to: '{self.expected_current_channel}'"
+            )
```

Comment:
I've thought about this some more and came around it making sense for this code to be here. BootStack also confirmed they would appreciate automating cs → ch migrations.
I would however not call the function `_get_refresh_current_channel_plan` since it may yield a channel switch. How about `_get_refresh_current_release_charm_plan` or even simply `_get_refresh_charm_plan`? (and then `_get_refresh_next_channel_plan` becomes `_get_upgrade_charm_plan`)

Finally, we should think about what should happen when you encounter an application running on channel `ussuri/edge`. Should we automatically switch to `ussuri/stable`?

Reply:
I've changed the naming:

-  `_get_refresh_current_channel_plan` to `_get_refresh_charm_plan`
- `_get_refresh_next_channel_plan` to `_get_upgrade_charm_plan`

I liked this naming idea. Thanks

> Finally, we should think about what should happen when you encounter an application running on channel ussuri/edge. Should we automatically switch to ussuri/stable?

From the [docs](https://docs.openstack.org/charm-guide/latest/admin/upgrades/openstack.html#manual-intervention):
`By design, the latest stable charms will support the software changes related to the OpenStack services being upgraded.`

So, in theory we should use `stable`, but it might have the case of a bug-fix that not landed yet. We are showing that we are going to change from `x` to `y`. I think this is a choice from the operator to continue or not.